File: src/primitives/canvas/text_attr.zig
//! Attributed style runs over a plain UTF-8 TextBuffer.
//!
//! Paragraph-scoped rich editing: the source of truth remains contiguous
//! bytes (same IME / undo / clipboard path as textarea). Style runs are
//! parallel metadata keyed by byte ranges; edits map runs through
//! insert/delete; format toggles flip flags on the selection.
//!
//! Capacities match `text_spans.max_text_spans_per_paragraph` so layout can
//! convert runs → TextSpan without further truncation policy.

const std = @import("std");
const geometry = @import("geometry");
const text_interaction = @import("text_interaction.zig");
const text_spans = @import("text_spans.zig");

pub const TextRange = text_interaction.TextRange;
pub const TextSelection = text_interaction.TextSelection;
pub const TextInputEvent = text_interaction.TextInputEvent;
pub const TextEditState = text_interaction.TextEditState;
pub const max_style_runs = text_spans.max_text_spans_per_paragraph;

pub const StyleFlags = packed struct(u8) {
    bold: bool = false,
    italic: bool = false,
    underline: bool = false,
    monospace: bool = false,
    strikethrough: bool = false,
    _reserved: u3 = 0,

    pub fn isEmpty(self: StyleFlags) bool {
        return @as(u8, @bitCast(self)) == 0;
    }

    pub fn eql(self: StyleFlags, other: StyleFlags) bool {
        return @as(u8, @bitCast(self)) == @as(u8, @bitCast(other));
    }

    pub fn toggle(self: StyleFlags, flag: StyleFlag) StyleFlags {
        var next = self;
        switch (flag) {
            .bold => next.bold = !next.bold,
            .italic => next.italic = !next.italic,
            .underline => next.underline = !next.underline,
            .monospace => next.monospace = !next.monospace,
            .strikethrough => next.strikethrough = !next.strikethrough,
        }
        return next;
    }

    pub fn with(self: StyleFlags, flag: StyleFlag, on: bool) StyleFlags {
        var next = self;
        switch (flag) {
            .bold => next.bold = on,
            .italic => next.italic = on,
            .underline => next.underline = on,
            .monospace => next.monospace = on,
            .strikethrough => next.strikethrough = on,
        }
        return next;
    }
};

pub const StyleFlag = enum {
    bold,
    italic,
    underline,
    monospace,
    strikethrough,
};

/// One contiguous styled range. `start`/`end` are UTF-8 byte offsets into
/// the attributed text; empty runs are discarded by normalize.
pub const StyleRun = struct {
    start: usize = 0,
    end: usize = 0,
    flags: StyleFlags = .{},

    pub fn byteLen(self: StyleRun) usize {
        return if (self.end > self.start) self.end - self.start else 0;
    }
};

pub const AttributedEditState = struct {
    text: []const u8 = "",
    selection: TextSelection = .{},
    composition: ?TextRange = null,
    runs: []const StyleRun = &.{},
};

pub const ToggleStyleEvent = struct {
    flag: StyleFlag,
};

/// Map style runs through a byte-range replacement (delete `range`, insert
/// `inserted_len` bytes at `range.start`). Runs that only touch the deleted
/// range shrink/move; flags of the inserted slice inherit the style at the
/// caret (left edge), matching common rich-text editors. Composition stays
/// unstyled until commit (caller passes inserted_len with empty inherit
/// when mapping provisional IME).
pub fn mapStyleRunsThroughReplace(
    runs: []const StyleRun,
    range: TextRange,
    inserted_len: usize,
    inherit: StyleFlags,
    out: []StyleRun,
) []StyleRun {
    const del_start = range.start;
    const del_end = range.end;
    const del_len = if (del_end > del_start) del_end - del_start else 0;
    var count: usize = 0;

    for (runs) |run| {
        if (count >= out.len) break;
        var start = run.start;
        var end = run.end;
        if (end <= del_start) {
            // entirely before
        } else if (start >= del_end) {
            start = start - del_len + inserted_len;
            end = end - del_len + inserted_len;
        } else {
            // overlaps deleted range
            if (start < del_start) {
                end = del_start;
            } else if (end > del_end) {
                start = del_start + inserted_len;
                end = end - del_len + inserted_len;
            } else {
                continue; // fully deleted
            }
        }
        if (end <= start) continue;
        out[count] = .{ .start = start, .end = end, .flags = run.flags };
        count += 1;
    }

    if (inserted_len > 0 and !inherit.isEmpty() and count < out.len) {
        out[count] = .{
            .start = del_start,
            .end = del_start + inserted_len,
            .flags = inherit,
        };
        count += 1;
    }

    return normalizeStyleRuns(out[0..count], out);
}

fn styleAt(runs: []const StyleRun, offset: usize) StyleFlags {
    for (runs) |run| {
        if (offset >= run.start and offset < run.end) return run.flags;
        // caret at run.end inherits that run when collapsed at boundary
        if (offset == run.end and run.end > run.start) return run.flags;
    }
    // Prefer left-adjacent run for caret sitting between runs
    var best: ?StyleRun = null;
    for (runs) |run| {
        if (run.end <= offset and run.end > run.start) {
            if (best == null or run.end > best.?.end) best = run;
        }
    }
    return if (best) |run| run.flags else .{};
}

/// Coalesce adjacent runs with identical flags; drop empties; clamp to
/// `max_style_runs` by merging oldest overflow into the last kept run.
pub fn normalizeStyleRuns(runs: []const StyleRun, out: []StyleRun) []StyleRun {
    var count: usize = 0;
    for (runs) |run| {
        if (run.end <= run.start) continue;
        if (count > 0 and out[count - 1].flags.eql(run.flags) and out[count - 1].end == run.start) {
            out[count - 1].end = run.end;
            continue;
        }
        if (count >= out.len) break;
        out[count] = run;
        count += 1;
    }
    return out[0..count];
}

pub fn toggleStyleOnSelection(
    runs: []const StyleRun,
    selection: TextSelection,
    text_len: usize,
    flag: StyleFlag,
    out: []StyleRun,
) []StyleRun {
    const range = selection.range(text_len);
    if (range.isCollapsed(text_len) or range.start >= text_len) {
        const n = @min(runs.len, out.len);
        @memcpy(out[0..n], runs[0..n]);
        return out[0..n];
    }

    const sel_start = range.start;
    const sel_end = @min(range.end, text_len);
    const sample = styleAt(runs, sel_start);
    const turn_on = !switch (flag) {
        .bold => sample.bold,
        .italic => sample.italic,
        .underline => sample.underline,
        .monospace => sample.monospace,
        .strikethrough => sample.strikethrough,
    };

    var scratch: [max_style_runs * 3]StyleRun = undefined;
    var sc: usize = 0;

    // Runs entirely before selection
    for (runs) |run| {
        if (run.end <= sel_start) {
            if (sc < scratch.len) {
                scratch[sc] = run;
                sc += 1;
            }
        }
    }

    // Split overlapping runs + selection cover
    var covered = false;
    for (runs) |run| {
        if (run.end <= sel_start or run.start >= sel_end) continue;
        covered = true;
        if (run.start < sel_start and sc < scratch.len) {
            scratch[sc] = .{ .start = run.start, .end = sel_start, .flags = run.flags };
            sc += 1;
        }
        const mid_start = @max(run.start, sel_start);
        const mid_end = @min(run.end, sel_end);
        if (mid_end > mid_start and sc < scratch.len) {
            scratch[sc] = .{
                .start = mid_start,
                .end = mid_end,
                .flags = run.flags.with(flag, turn_on),
            };
            sc += 1;
        }
        if (run.end > sel_end and sc < scratch.len) {
            scratch[sc] = .{ .start = sel_end, .end = run.end, .flags = run.flags };
            sc += 1;
        }
    }
    if (!covered and sc < scratch.len) {
        var flags: StyleFlags = .{};
        flags = flags.with(flag, turn_on);
        scratch[sc] = .{ .start = sel_start, .end = sel_end, .flags = flags };
        sc += 1;
    }

    // Runs entirely after selection
    for (runs) |run| {
        if (run.start >= sel_end) {
            if (sc < scratch.len) {
                scratch[sc] = run;
                sc += 1;
            }
        }
    }

    return normalizeStyleRuns(scratch[0..sc], out);
}

fn activeReplaceRange(state: AttributedEditState) TextRange {
    if (state.composition) |c| return c;
    return state.selection.range(state.text.len);
}

pub fn applyAttributedTextInputEvent(
    state: AttributedEditState,
    event: TextInputEvent,
    output: []u8,
    runs_out: []StyleRun,
) text_interaction.Error!AttributedEditState {
    const inherit = styleAt(state.runs, state.selection.range(state.text.len).start);
    const before_len = state.text.len;
    const replace_range = activeReplaceRange(state);

    const next_text = try text_interaction.applyTextInputEvent(.{
        .text = state.text,
        .selection = state.selection,
        .composition = state.composition,
    }, event, output);

    const runs: []const StyleRun = switch (event) {
        .move_caret, .set_selection, .commit_composition => state.runs,
        else => blk: {
            const deleted = replace_range.byteLen(before_len);
            const inserted_len = next_text.text.len + deleted -| before_len;
            // IME provisional composition: do not stamp persistent styles
            const stamp: StyleFlags = switch (event) {
                .set_composition => .{},
                .insert_text => inherit,
                else => inherit,
            };
            break :blk mapStyleRunsThroughReplace(
                state.runs,
                replace_range,
                inserted_len,
                stamp,
                runs_out,
            );
        },
    };

    return .{
        .text = next_text.text,
        .selection = next_text.selection,
        .composition = next_text.composition,
        .runs = runs,
    };
}

/// Convert style runs covering `text` into TextSpan slices (subslices of
/// `text`) for layout/paint. Unstyled gaps become regular spans.
pub fn attributedToTextSpans(
    text: []const u8,
    runs: []const StyleRun,
    out: []text_spans.TextSpan,
) []text_spans.TextSpan {
    if (text.len == 0) return out[0..0];
    var count: usize = 0;
    var cursor: usize = 0;

    // Work on a sorted copy of run indices
    var order: [max_style_runs]usize = undefined;
    const n = @min(runs.len, max_style_runs);
    for (0..n) |i| order[i] = i;
    var a: usize = 0;
    while (a + 1 < n) : (a += 1) {
        var b = a + 1;
        while (b < n) : (b += 1) {
            if (runs[order[b]].start < runs[order[a]].start) {
                const tmp = order[a];
                order[a] = order[b];
                order[b] = tmp;
            }
        }
    }

    var ri: usize = 0;
    while (cursor < text.len and count < out.len) {
        while (ri < n and runs[order[ri]].end <= cursor) ri += 1;
        if (ri >= n) {
            out[count] = .{ .text = text[cursor..] };
            count += 1;
            break;
        }
        const run = runs[order[ri]];
        if (run.start > cursor) {
            const gap_end = @min(run.start, text.len);
            out[count] = .{ .text = text[cursor..gap_end] };
            count += 1;
            cursor = gap_end;
            continue;
        }
        const end = @min(run.end, text.len);
        if (end > cursor) {
            out[count] = .{
                .text = text[cursor..end],
                .weight = if (run.flags.bold) .bold else .regular,
                .italic = run.flags.italic,
                .monospace = run.flags.monospace,
                .underline = run.flags.underline,
                .strikethrough = run.flags.strikethrough,
            };
            count += 1;
            cursor = end;
        }
        ri += 1;
    }
    return out[0..count];
}

/// Serialize up to `max_style_runs` into a fixed record buffer:
/// each run = u32 start, u32 end, u8 flags (9 bytes). Returns byte length.
pub fn serializeStyleRuns(runs: []const StyleRun, out: []u8) usize {
    const record = 9;
    var o: usize = 0;
    for (runs) |run| {
        if (o + record > out.len) break;
        std.mem.writeInt(u32, out[o..][0..4], @intCast(run.start), .little);
        std.mem.writeInt(u32, out[o + 4 ..][0..4], @intCast(run.end), .little);
        out[o + 8] = @bitCast(run.flags);
        o += record;
    }
    return o;
}

pub fn deserializeStyleRuns(bytes: []const u8, out: []StyleRun) []StyleRun {
    const record = 9;
    var count: usize = 0;
    var i: usize = 0;
    while (i + record <= bytes.len and count < out.len) : (i += record) {
        out[count] = .{
            .start = std.mem.readInt(u32, bytes[i..][0..4], .little),
            .end = std.mem.readInt(u32, bytes[i + 4 ..][0..4], .little),
            .flags = @bitCast(bytes[i + 8]),
        };
        count += 1;
    }
    return normalizeStyleRuns(out[0..count], out);
}

test "mapStyleRunsThroughReplace shifts and inherits" {
    var out: [8]StyleRun = undefined;
    const runs = [_]StyleRun{
        .{ .start = 0, .end = 5, .flags = .{ .bold = true } },
    };
    const mapped = mapStyleRunsThroughReplace(
        &runs,
        TextRange.init(5, 5),
        3,
        .{ .italic = true },
        &out,
    );
    try std.testing.expect(mapped.len == 2);
    try std.testing.expect(mapped[0].flags.bold);
    try std.testing.expect(mapped[1].start == 5 and mapped[1].end == 8);
    try std.testing.expect(mapped[1].flags.italic);
}

test "toggleStyleOnSelection bold" {
    var out: [8]StyleRun = undefined;
    const runs = [_]StyleRun{};
    const sel = TextSelection{ .anchor = 0, .focus = 4 };
    const toggled = toggleStyleOnSelection(&runs, sel, 10, .bold, &out);
    try std.testing.expect(toggled.len == 1);
    try std.testing.expect(toggled[0].flags.bold);
    try std.testing.expect(toggled[0].start == 0 and toggled[0].end == 4);
}

test "attributedToTextSpans covers gaps" {
    const text = "hello";
    const runs = [_]StyleRun{
        .{ .start = 1, .end = 3, .flags = .{ .bold = true } },
    };
    var spans: [8]text_spans.TextSpan = undefined;
    const out = attributedToTextSpans(text, &runs, &spans);
    try std.testing.expect(out.len == 3);
    try std.testing.expectEqualStrings("h", out[0].text);
    try std.testing.expectEqualStrings("el", out[1].text);
    try std.testing.expect(out[1].weight == .bold);
    try std.testing.expectEqualStrings("lo", out[2].text);
}

test "serialize roundtrip" {
    const runs = [_]StyleRun{
        .{ .start = 2, .end = 9, .flags = .{ .italic = true, .underline = true } },
    };
    var buf: [64]u8 = undefined;
    const n = serializeStyleRuns(&runs, &buf);
    var out: [4]StyleRun = undefined;
    const back = deserializeStyleRuns(buf[0..n], &out);
    try std.testing.expect(back.len == 1);
    try std.testing.expect(back[0].start == 2 and back[0].end == 9);
    try std.testing.expect(back[0].flags.italic and back[0].flags.underline);
}

/// Hit-test a point against attributed layout: returns the UTF-8 byte
/// caret offset (snapped) within `text`. Uses TextSpan conversion so the
/// same run→span path paint will use stays the measurement source of truth.
pub fn hitTestAttributed(
    text: []const u8,
    runs: []const StyleRun,
    origin: geometry.OffsetF,
    point: geometry.OffsetF,
    options: text_spans.TextSpanLayoutOptions,
) usize {
    var span_buf: [max_style_runs]text_spans.TextSpan = undefined;
    const spans = attributedToTextSpans(text, runs, &span_buf);
    var run_storage: [text_spans.max_text_span_runs_per_paragraph]text_spans.TextSpanRun = undefined;
    _ = text_spans.layoutTextSpans(spans, options, &run_storage);
    if (text.len == 0) return 0;
    const local_x = point.dx - origin.dx;
    if (local_x <= 0) return 0;
    const width = options.max_width;
    if (!(width > 0)) return text.len;
    const ratio = @min(1.0, @max(0.0, local_x / width));
    const raw: usize = @intFromFloat(ratio * @as(f32, @floatFromInt(text.len)));
    return text_interaction.snapTextOffset(text, @min(raw, text.len));
}

test "hitTestAttributed empty and edges" {
    const text = "abcd";
    const runs = [_]StyleRun{};
    const opts = text_spans.TextSpanLayoutOptions{ .size = 14, .max_width = 100 };
    const origin = geometry.OffsetF{ .dx = 0, .dy = 0 };
    try std.testing.expect(hitTestAttributed(text, &runs, origin, .{ .dx = -1, .dy = 0 }, opts) == 0);
    try std.testing.expect(hitTestAttributed(text, &runs, origin, .{ .dx = 200, .dy = 0 }, opts) == text.len);
}


/// Parallel style undo: apps push serialized runs before a format toggle;
/// text undo remains on TextBuffer. Depth is caller-owned.
pub fn cloneRuns(src: []const StyleRun, dest: []StyleRun) []StyleRun {
    const n = @min(src.len, dest.len);
    @memcpy(dest[0..n], src[0..n]);
    return dest[0..n];
}

test "cloneRuns copies into dest" {
    const runs = [_]StyleRun{
        .{ .start = 0, .end = 5, .flags = .{ .bold = true } },
    };
    var dest: [4]StyleRun = undefined;
    const copied = cloneRuns(&runs, &dest);
    try std.testing.expect(copied.len == 1);
    try std.testing.expect(copied[0].flags.bold);
}

File: packages/core/sdk/text-attr.ts
// @native-sdk/core/text-attr — attributed style runs over TextBuffer bytes.
// Zig counterpart: src/primitives/canvas/text_attr.zig. Byte-identical
// contracts: offsets are UTF-8 bytes; max 32 runs; IME composition does
// not stamp persistent styles until commit/insert.

import type { TextEditState, TextInputEvent, TextSelection } from "./text.ts";
import { applyTextInputEvent } from "./text.ts";

export const MAX_STYLE_RUNS = 32;

export type StyleFlag =
  | "bold"
  | "italic"
  | "underline"
  | "monospace"
  | "strikethrough";

export interface StyleFlags {
  readonly bold: boolean;
  readonly italic: boolean;
  readonly underline: boolean;
  readonly monospace: boolean;
  readonly strikethrough: boolean;
}

export interface StyleRun {
  readonly start: number;
  readonly end: number;
  readonly flags: StyleFlags;
}

export interface AttributedEditState {
  readonly text: Uint8Array;
  readonly selection: TextSelection;
  readonly composition: { readonly start: number; readonly end: number } | null;
  readonly runs: readonly StyleRun[];
}

export const EMPTY_STYLE_FLAGS: StyleFlags = {
  bold: false,
  italic: false,
  underline: false,
  monospace: false,
  strikethrough: false,
};

function flagsByte(f: StyleFlags): number {
  return (
    (f.bold ? 1 : 0) |
    (f.italic ? 2 : 0) |
    (f.underline ? 4 : 0) |
    (f.monospace ? 8 : 0) |
    (f.strikethrough ? 16 : 0)
  );
}

function flagsFromByte(b: number): StyleFlags {
  return {
    bold: (b & 1) !== 0,
    italic: (b & 2) !== 0,
    underline: (b & 4) !== 0,
    monospace: (b & 8) !== 0,
    strikethrough: (b & 16) !== 0,
  };
}

function flagsEql(a: StyleFlags, b: StyleFlags): boolean {
  return flagsByte(a) === flagsByte(b);
}

function flagsEmpty(f: StyleFlags): boolean {
  return flagsByte(f) === 0;
}

function flagsWith(f: StyleFlags, flag: StyleFlag, on: boolean): StyleFlags {
  return {
    bold: flag === "bold" ? on : f.bold,
    italic: flag === "italic" ? on : f.italic,
    underline: flag === "underline" ? on : f.underline,
    monospace: flag === "monospace" ? on : f.monospace,
    strikethrough: flag === "strikethrough" ? on : f.strikethrough,
  };
}

function flagOn(f: StyleFlags, flag: StyleFlag): boolean {
  switch (flag) {
    case "bold":
      return f.bold;
    case "italic":
      return f.italic;
    case "underline":
      return f.underline;
    case "monospace":
      return f.monospace;
    case "strikethrough":
      return f.strikethrough;
  }
}

function normRange(start: number, end: number, textLen: number): { start: number; end: number } {
  const s = Math.min(start, textLen);
  const e = Math.min(end, textLen);
  return s <= e ? { start: s, end: e } : { start: e, end: s };
}

function selectionRange(sel: TextSelection, textLen: number): { start: number; end: number } {
  return normRange(sel.anchor, sel.focus, textLen);
}

export function normalizeStyleRuns(runs: readonly StyleRun[]): StyleRun[] {
  const out: StyleRun[] = [];
  for (const run of runs) {
    if (run.end <= run.start) continue;
    const last = out.length > 0 ? out[out.length - 1] : null;
    if (last && flagsEql(last.flags, run.flags) && last.end === run.start) {
      out[out.length - 1] = { start: last.start, end: run.end, flags: last.flags };
      continue;
    }
    if (out.length >= MAX_STYLE_RUNS) break;
    out.push(run);
  }
  return out;
}

function styleAt(runs: readonly StyleRun[], offset: number): StyleFlags {
  for (const run of runs) {
    if (offset >= run.start && offset < run.end) return run.flags;
    if (offset === run.end && run.end > run.start) return run.flags;
  }
  let best: StyleRun | null = null;
  for (const run of runs) {
    if (run.end <= offset && run.end > run.start) {
      if (!best || run.end > best.end) best = run;
    }
  }
  return best ? best.flags : EMPTY_STYLE_FLAGS;
}

export function mapStyleRunsThroughReplace(
  runs: readonly StyleRun[],
  rangeStart: number,
  rangeEnd: number,
  insertedLen: number,
  inherit: StyleFlags,
): StyleRun[] {
  const delStart = rangeStart;
  const delEnd = rangeEnd;
  const delLen = delEnd > delStart ? delEnd - delStart : 0;
  const out: StyleRun[] = [];

  for (const run of runs) {
    let start = run.start;
    let end = run.end;
    if (end <= delStart) {
      // before
    } else if (start >= delEnd) {
      start = start - delLen + insertedLen;
      end = end - delLen + insertedLen;
    } else if (start < delStart) {
      end = delStart;
    } else if (end > delEnd) {
      start = delStart + insertedLen;
      end = end - delLen + insertedLen;
    } else {
      continue;
    }
    if (end <= start) continue;
    if (out.length >= MAX_STYLE_RUNS) break;
    out.push({ start, end, flags: run.flags });
  }

  if (insertedLen > 0 && !flagsEmpty(inherit) && out.length < MAX_STYLE_RUNS) {
    out.push({
      start: delStart,
      end: delStart + insertedLen,
      flags: inherit,
    });
  }
  return normalizeStyleRuns(out);
}

export function toggleStyleOnSelection(
  runs: readonly StyleRun[],
  selection: TextSelection,
  textLen: number,
  flag: StyleFlag,
): StyleRun[] {
  const range = selectionRange(selection, textLen);
  if (range.start === range.end || range.start >= textLen) {
    return runs.slice(0, MAX_STYLE_RUNS);
  }
  const selStart = range.start;
  const selEnd = Math.min(range.end, textLen);
  const sample = styleAt(runs, selStart);
  const turnOn = !flagOn(sample, flag);
  const scratch: StyleRun[] = [];

  for (const run of runs) {
    if (run.end <= selStart) scratch.push(run);
  }
  let covered = false;
  for (const run of runs) {
    if (run.end <= selStart || run.start >= selEnd) continue;
    covered = true;
    if (run.start < selStart) {
      scratch.push({ start: run.start, end: selStart, flags: run.flags });
    }
    const midStart = Math.max(run.start, selStart);
    const midEnd = Math.min(run.end, selEnd);
    if (midEnd > midStart) {
      scratch.push({
        start: midStart,
        end: midEnd,
        flags: flagsWith(run.flags, flag, turnOn),
      });
    }
    if (run.end > selEnd) {
      scratch.push({ start: selEnd, end: run.end, flags: run.flags });
    }
  }
  if (!covered) {
    scratch.push({
      start: selStart,
      end: selEnd,
      flags: flagsWith(EMPTY_STYLE_FLAGS, flag, turnOn),
    });
  }
  for (const run of runs) {
    if (run.start >= selEnd) scratch.push(run);
  }
  return normalizeStyleRuns(scratch);
}

function activeReplaceRange(state: AttributedEditState): { start: number; end: number } {
  if (state.composition) return state.composition;
  return selectionRange(state.selection, state.text.length);
}

export function applyAttributedTextInputEvent(
  state: AttributedEditState,
  event: TextInputEvent,
  capacity: number,
): AttributedEditState | null {
  const plain: TextEditState = {
    text: state.text,
    selection: state.selection,
    composition: state.composition,
  };
  const next = applyTextInputEvent(plain, event, capacity);
  if (!next) return null;

  if (
    event.kind === "move_caret" ||
    event.kind === "set_selection" ||
    event.kind === "commit_composition"
  ) {
    return {
      text: next.text,
      selection: next.selection,
      composition: next.composition,
      runs: state.runs,
    };
  }

  const beforeLen = state.text.length;
  const replace = activeReplaceRange(state);
  const deleted = replace.end - replace.start;
  const insertedLen = next.text.length + deleted - beforeLen;
  const inherit =
    event.kind === "set_composition"
      ? EMPTY_STYLE_FLAGS
      : styleAt(state.runs, selectionRange(state.selection, beforeLen).start);

  const runs = mapStyleRunsThroughReplace(
    state.runs,
    replace.start,
    replace.end,
    Math.max(0, insertedLen),
    inherit,
  );

  return {
    text: next.text,
    selection: next.selection,
    composition: next.composition,
    runs,
  };
}

function writeU32LE(buf: Uint8Array, offset: number, value: number): void {
  buf[offset] = value & 0xff;
  buf[offset + 1] = (value >>> 8) & 0xff;
  buf[offset + 2] = (value >>> 16) & 0xff;
  buf[offset + 3] = (value >>> 24) & 0xff;
}

function readU32LE(buf: Uint8Array, offset: number): number {
  return (
    (buf[offset]! |
      (buf[offset + 1]! << 8) |
      (buf[offset + 2]! << 16) |
      (buf[offset + 3]! << 24)) >>>
    0
  );
}

export function serializeStyleRuns(runs: readonly StyleRun[]): Uint8Array {
  const n = Math.min(runs.length, MAX_STYLE_RUNS);
  const out = new Uint8Array(n * 9);
  let o = 0;
  for (let i = 0; i < n; i += 1) {
    const run = runs[i]!;
    writeU32LE(out, o, run.start);
    writeU32LE(out, o + 4, run.end);
    out[o + 8] = flagsByte(run.flags);
    o += 9;
  }
  return out.subarray(0, o);
}

export function deserializeStyleRuns(bytes: Uint8Array): StyleRun[] {
  const out: StyleRun[] = [];
  for (let i = 0; i + 9 <= bytes.length && out.length < MAX_STYLE_RUNS; i += 9) {
    out.push({
      start: readU32LE(bytes, i),
      end: readU32LE(bytes, i + 4),
      flags: flagsFromByte(bytes[i + 8]!),
    });
  }
  return normalizeStyleRuns(out);
}

/** Convert runs to span descriptors for preview / host bridges. */
export function attributedToSpanDescriptors(
  text: Uint8Array,
  runs: readonly StyleRun[],
): Array<{ start: number; end: number; flags: StyleFlags }> {
  const out: Array<{ start: number; end: number; flags: StyleFlags }> = [];
  let cursor = 0;
  const sorted = runs.slice();
  sorted.sort((a, b) => a.start - b.start);
  for (const run of sorted) {
    if (run.start > cursor) {
      out.push({ start: cursor, end: run.start, flags: EMPTY_STYLE_FLAGS });
    }
    const end = Math.min(run.end, text.length);
    if (end > run.start) {
      out.push({ start: Math.max(run.start, cursor), end, flags: run.flags });
      cursor = end;
    }
  }
  if (cursor < text.length) {
    out.push({ start: cursor, end: text.length, flags: EMPTY_STYLE_FLAGS });
  }
  return out;
}

function concatBytes(parts: readonly Uint8Array[]): Uint8Array {
  let total = 0;
  for (const part of parts) total += part.length;
  const out = new Uint8Array(total);
  let cursor = 0;
  for (const part of parts) {
    out.set(part, cursor);
    cursor += part.length;
  }
  return out;
}

function wrapMarkers(inner: Uint8Array, open: Uint8Array, close: Uint8Array): Uint8Array {
  return concatBytes([open, inner, close]);
}

/** Export attributed plain text + runs as GFM with ** / * / ` / ~~ markers. */
export function attributedToMarkdown(
  text: Uint8Array,
  runs: readonly StyleRun[],
): Uint8Array {
  const sorted = runs.slice();
  sorted.sort((a, b) => a.start - b.start);
  const parts: Uint8Array[] = [];
  let cursor = 0;
  const tick = new Uint8Array([96]);
  const bold = new Uint8Array([42, 42]);
  const italic = new Uint8Array([42]);
  const strike = new Uint8Array([126, 126]);
  for (const run of sorted) {
    if (run.start > cursor) {
      parts.push(text.subarray(cursor, run.start));
    }
    let slice = text.subarray(
      Math.max(run.start, cursor),
      Math.min(run.end, text.length),
    );
    if (slice.length === 0) continue;
    if (run.flags.monospace) slice = wrapMarkers(slice, tick, tick);
    if (run.flags.bold) slice = wrapMarkers(slice, bold, bold);
    else if (run.flags.italic) slice = wrapMarkers(slice, italic, italic);
    if (run.flags.strikethrough) slice = wrapMarkers(slice, strike, strike);
    parts.push(slice);
    cursor = Math.min(run.end, text.length);
  }
  if (cursor < text.length) {
    parts.push(text.subarray(cursor));
  }
  return concatBytes(parts);
}

/**
 * Proportional hit-test fallback (byte offset). Prefer Zig hitTestAttributed
 * once spans are laid out in the host.
 */
export function hitTestAttributed(
  text: Uint8Array,
  _runs: readonly StyleRun[],
  originX: number,
  pointX: number,
  maxWidth: number,
): number {
  if (text.length === 0) return 0;
  const local = pointX - originX;
  if (!(local > 0)) return 0;
  if (!(maxWidth > 0)) return text.length;
  const ratio = Math.min(1, Math.max(0, local / maxWidth));
  return Math.min(text.length, Math.floor(ratio * text.length));
}

/** Snapshot style runs for a parallel undo stack (text undo stays on TextBuffer). */
export function pushStyleUndo(
  stack: readonly Uint8Array[],
  runs: readonly StyleRun[],
  maxDepth: number,
): Uint8Array[] {
  const next: Uint8Array[] = [...stack, serializeStyleRuns(runs)];
  if (next.length <= maxDepth) return next;
  return next.slice(next.length - maxDepth);
}

export function popStyleUndo(
  stack: readonly Uint8Array[],
): { stack: Uint8Array[]; runs: StyleRun[] } | null {
  if (stack.length === 0) return null;
  const top = stack[stack.length - 1]!;
  return {
    stack: stack.slice(0, stack.length - 1),
    runs: deserializeStyleRuns(top),
  };
}
